// danger/distribution/platforms/LocalGit.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var gitJSONToGitDSL_1 = require("./git/gitJSONToGitDSL");
var diffToGitJSONDSL_1 = require("./git/diffToGitJSONDSL");
var localGetDiff_1 = require("./git/localGetDiff");
var localGetFileAtSHA_1 = require("./git/localGetFileAtSHA");
var localGetCommits_1 = require("./git/localGetCommits");
var fs_1 = require("fs");
var LocalGit = /** @class */ (function () {
    function LocalGit(options) {
        this.options = options;
        this.getFileContents = function (path) { return new Promise(function (res) { return res(fs_1.readFileSync(path, "utf8")); }); };
        this.name = "local git";
    }
    LocalGit.prototype.getPlatformDSLRepresentation = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, null];
            });
        });
    };
    LocalGit.prototype.getPlatformGitRepresentation = function () {
        return __awaiter(this, void 0, void 0, function () {
            var base, head, diff, commits, gitJSON, config;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        base = this.options.base || "master";
                        head = "HEAD";
                        return [4 /*yield*/, localGetDiff_1.localGetDiff(base, head)];
                    case 1:
                        diff = _a.sent();
                        return [4 /*yield*/, localGetCommits_1.localGetCommits(base, head)];
                    case 2:
                        commits = _a.sent();
                        gitJSON = diffToGitJSONDSL_1.diffToGitJSONDSL(diff, commits);
                        config = {
                            repo: process.cwd(),
                            baseSHA: this.options.base || "master",
                            headSHA: "HEAD",
                            getFileContents: localGetFileAtSHA_1.localGetFileAtSHA,
                            getFullDiff: localGetDiff_1.localGetDiff,
                        };
                        return [2 /*return*/, gitJSONToGitDSL_1.gitJSONToGitDSL(gitJSON, config)];
                }
            });
        });
    };
    LocalGit.prototype.supportsCommenting = function () {
        return false;
    };
    LocalGit.prototype.updateOrCreateComment = function (_newComment) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, true];
            });
        });
    };
    LocalGit.prototype.createComment = function (_comment) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, true];
            });
        });
    };
    LocalGit.prototype.deleteMainComment = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, true];
            });
        });
    };
    LocalGit.prototype.editMainComment = function (_comment) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, true];
            });
        });
    };
    LocalGit.prototype.updateStatus = function (_success, _message) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, true];
            });
        });
    };
    return LocalGit;
}());
exports.LocalGit = LocalGit;
