// danger/distribution/runner/templates/githubIssueTemplate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var v = require("voca");
/**
 * Converts a set of violations into a HTML table
 *
 * @param {string} name User facing title of table
 * @param {string} emoji Emoji name to show next to each item
 * @param {Violation[]} violations for table
 * @returns {string} HTML
 */
function table(name, emoji, violations) {
    if (violations.length === 0 || violations.every(function (violation) { return !violation.message; })) {
        return "";
    }
    return "\n<table>\n  <thead>\n    <tr>\n      <th width=\"50\"></th>\n      <th width=\"100%\" data-danger-table=\"true\">" + name + "</th>\n    </tr>\n  </thead>\n  <tbody>" + violations
        .map(function (v) {
        return "<tr>\n      <td>:" + emoji + ":</td>\n      <td>\n\n  " + v.message + "\n  </td>\n    </tr>\n  ";
    })
        .join("\n") + "</tbody>\n</table>\n";
}
function getSummary(label, violations) {
    return violations
        .map(function (x) { return v.truncate(x.message, 20); })
        .reduce(function (acc, value, idx) { return acc + " " + value + (idx === violations.length - 1 ? "" : ","); }, violations.length + " " + label + ": ");
}
function buildSummaryMessage(dangerID, results) {
    var fails = results.fails, warnings = results.warnings, messages = results.messages, markdowns = results.markdowns;
    var summary = "  " + getSummary("failure", fails) + "\n  " + getSummary("warning", warnings) + "\n  " + (messages.length > 0 ? messages.length + " messages" : "") + "\n  " + (markdowns.length > 0 ? markdowns.length + " markdown notices" : "") + "\n  " + exports.dangerIDToString(dangerID);
    return summary;
}
exports.dangerIDToString = function (id) { return "DangerID: danger-id-" + id + ";"; };
/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
exports.dangerSignaturePostfix = "Generated by :no_entry_sign: <a href=\"http://github.com/danger/danger-js/\">dangerJS</a>";
/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {DangerResults} results Data to work with
 * @returns {string} HTML
 */
function template(dangerID, results) {
    return "\n<!--\n" + buildSummaryMessage(dangerID, results) + "\n-->\n" + table("Fails", "no_entry_sign", results.fails) + "\n" + table("Warnings", "warning", results.warnings) + "\n" + table("Messages", "book", results.messages) + "\n" + results.markdowns.join("\n\n") + "\n<p align=\"right\">\n  " + exports.dangerSignaturePostfix + "\n</p>\n";
}
exports.template = template;
